pkg/provider/cloud/openstack/provider.go

```diff
@@ -195,233 +199,377 @@ func validateExistingSubnetOverlap(networkID string, netClient *gophercloud.Serv
+		network, err := getNetworkByName(netClient, cluster.Spec.Cloud.Openstack.Network, false)
 		if err != nil {
-			return nil, fmt.Errorf("failed to verify that the subnet '%s' has a router attached: %w", cluster.Spec.Cloud.Openstack.IPv6SubnetID, err)
+			return nil, fmt.Errorf("failed to get network '%s': %w", cluster.Spec.Cloud.Openstack.Network, err)
+		}
+		cluster, err = reconcileIPv4Subnet(ctx, netClient, cluster, update, network.ID, os.dc.DNSServers)
+		if err != nil {
+			return nil, err
+		}
+		cluster, err = reconcileIPv6Subnet(ctx, netClient, cluster, update, network.ID, os.dc.DNSServers)
+		if err != nil {
+			return nil, err
 		}
 	}
-	// If router has to be created, add associated finalizer.
-	if cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "" {
-		finalizers = append(finalizers, RouterCleanupFinalizer)
+	if force || cluster.Spec.Cloud.Openstack.RouterID == "" {
+		cluster, err = reconcileRouter(ctx, netClient, cluster, update)
+		if err != nil {
+			return nil, err
+		}
 	}
+	return cluster, nil
+}
 
-	if ipv4Network {
-		// If subnet has to be created, add associated finalizer.
-		if cluster.Spec.Cloud.Openstack.SubnetID == "" {
-			finalizers = append(finalizers, SubnetCleanupFinalizer)
+func reconcileNetwork(ctx context.Context, netClient *gophercloud.ServiceClient, cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
+	networkName := cluster.Spec.Cloud.Openstack.Network
+
+	// If the network name is set, check if it exists
+	if networkName != "" {
+		existingNetwork, err := getNetworkByName(netClient, networkName, false)
+		if err != nil {
+			if !isNotFoundErr(err) {
+				return nil, fmt.Errorf("failed to check for existing network: %w", err)
+			}
 		}
-		// If subnet or router has to be created, subnet needs to be attached to the router.
-		if cluster.Spec.Cloud.Openstack.SubnetID == "" || (cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "") {
-			finalizers = append(finalizers, RouterSubnetLinkCleanupFinalizer)
+		if existingNetwork != nil {
+			return cluster, nil
 		}
 	}
-	if ipv6Network {
-		// If subnet has to be created, add associated finalizer.
-		if cluster.Spec.Cloud.Openstack.IPv6SubnetID == "" {
-			finalizers = append(finalizers, IPv6SubnetCleanupFinalizer)
-		}
-		// If subnet or router has to be created, subnet needs to be attached to the router.
-		if cluster.Spec.Cloud.Openstack.IPv6SubnetID == "" || (cluster.Spec.Cloud.Openstack.RouterID == "" && routerID == "") {
-			finalizers = append(finalizers, RouterIPv6SubnetLinkCleanupFinalizer)
+
+	// If NetworkName not specified, Create network with name kubernetes-clusterid
+	if networkName == "" {
+		networkName = cluster.Name
+	}
+
+	newNetwork, err := createKubermaticNetwork(netClient, networkName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the kubermatic network: %w", err)
```

Comment:
This might be from the previous code, but why are we calling it "the kubermatic network"? Should this maybe be "the cluster network" instead?

Reply:
I didn't change the existing (older) helper functions name. If it's okay, I will go with it.

Reply:
Changed to `createUserClusterNetwork`